internal/server/filesystem/server.go

```diff
@@ -2,24 +2,227 @@ package filesystem
 
 import (
 	"context"
+	"fmt"
+	"regexp"
+	"strings"
 
-	"github.com/kubernetes-csi/csi-proxy/client/api"
 	"github.com/kubernetes-csi/csi-proxy/client/apiversion"
 	"github.com/kubernetes-csi/csi-proxy/internal/server/filesystem/internal"
 )
 
-// Server is the internal server for the filesytem API group.
-type Server struct{}
+type Server struct {
+	kubeletCSIPluginsPath string
+	kubeletPodPath        string
+	hostAPI               API
+}
+
+const MAX_PATH_LENGTH_WINDOWS = 260
+
+var invalidPathCharsRegexWindows = regexp.MustCompile(`["/\:\?\*|]`)
+
+type API interface {
+	PathExists(path string) (bool, error)
+	Mkdir(path string) error
+	Rmdir(path string) error
+	LinkPath(tgt string, src string) error
+}
+
+func NewServer(hostOS string, kubeletCSIPluginsPath string, kubeletPodPath string, hostAPI API) (*Server, error) {
+	if hostOS != "windows" {
+		return nil, fmt.Errorf("Unsupported OS for FileSystem API server: %s", hostOS)
+	}
+	return &Server{
+		kubeletCSIPluginsPath: kubeletCSIPluginsPath,
+		kubeletPodPath:        kubeletPodPath,
+		hostAPI:               hostAPI,
+	}, nil
+}
+
+func isDriveLetterWindows(c uint8) bool {
+	return ('a' <= c && c <= 'z' || 'A' <= c && c <= 'Z')
+}
+
+func containsInvalidCharactersWindows(path string) bool {
+	if isAbsWindows(path) {
+		path = path[3:]
+	}
+	if invalidPathCharsRegexWindows.MatchString(path) {
+		return true
+	}
+	if strings.Contains(path, `..`) {
+		return true
+	}
+	return false
+}
+
+func isUNCPathWindows(path string) bool {
+	// check for UNC/pipe prefixes like "\\"
+	if len(path) < 2 {
+		return false
+	}
+	if path[0] == '\\' && path[1] == '\\' {
+		return true
+	}
+	return false
+}
+
+func isAbsWindows(path string) bool {
+	// for Windows check for C:\\.. prefix only
+	// UNC prefixes of the form \\ are not considered
+	// absolute in the context of CSI proxy
+	if len(path) < 3 {
+		return false
+	}
+	c := path[0]
+	if !isDriveLetterWindows(c) {
+		return false
+	}
+	if path[1] != ':' {
+		return false
+	}
+	if path[2] != '\\' && path[2] != '/' {
+		return false
+	}
+	return true
+}
+
+func (s *Server) validatePathWindows(pathCtx internal.PathContext, path string) error {
+	prefix := ""
+	if pathCtx == internal.PLUGIN {
+		prefix = s.kubeletCSIPluginsPath
+	} else if pathCtx == internal.CONTAINER {
+		prefix = s.kubeletPodPath
+	} else {
+		return fmt.Errorf("Invalid PathContext: %v", pathCtx)
+	}
+
+	pathlen := len(path)
+
+	if pathlen > MAX_PATH_LENGTH_WINDOWS {
+		return fmt.Errorf("Path length %d exceeds maximum characters: %d", pathlen, MAX_PATH_LENGTH_WINDOWS)
+	}
+
+	if len(path) > 0 && (path[0] == '\\') {
```

Comment:
nit: we could use `pathlen` here instead of len(path)

Reply:
I don't see a `pathlen` builtin. Can you point me to the golang package that contains `pathlen` please @kkmsft

Reply:
Sorry! for the confusion. I was not talking about a package, but the variable being used in the codepath - pathlen  - instead of calculating len(path)